File: frontend/apps/web-antd/src/api/admin/plan.ts
/**
 * Plan management API / 套餐管理 API
 * Backend: /admin/plans/*
 */
import type { ApiRequestOptions } from '#/utils/request';

import { requestClient } from '#/utils/request';

// ============================================================
// Type definitions / 类型定义
// ============================================================

/** Billing cycle / 计费周期 */
export type BillingCycle =
  | 'custom'
  | 'lifetime'
  | 'monthly'
  | 'one_time'
  | 'quarterly'
  | 'yearly';

/** Quota config / 配额配置 */
export interface QuotaSchema {
  /** Allow custom domain / 是否允许自定义域名 */
  allowCustomDomain?: boolean | null;
  /** Monthly API call count / 每月 API 调用次数 */
  apiCallsPerMonth?: null | number;
  /** Max admins / 最大管理员数 */
  maxAdmins?: null | number;
  /** Max custom domains / 最大自定义域名数 */
  maxCustomDomains?: null | number;
  /** Max file size (MB) / 最大文件大小 */
  maxFileSizeMb?: null | number;
  /** Max users / 最大用户数 */
  maxUsers?: null | number;
  /** Storage limit (GB) / 存储限制 */
  storageLimitGb?: null | number;
}

/** Backend quota format (snake_case) / 后端配额格式 */
interface QuotaSchemaRaw {
  allow_custom_domain?: boolean | null;
  api_calls_per_month?: null | number;
  max_admins?: null | number;
  max_custom_domains?: null | number;
  max_file_size_mb?: null | number;
  max_users?: null | number;
  storage_limit_gb?: null | number;
}

/** Feature flags / 特性标记 */
export interface FeaturesSchema {
  /** Advanced analytics / 高级分析 */
  advancedAnalytics?: boolean | null;
  /** AI enabled / AI 功能 */
  aiEnabled?: boolean | null;
  /** Priority support / 优先支持 */
  prioritySupport?: boolean | null;
  /** Storage billing enabled / 对象存储账单对账收费 */
  storageBillingEnabled?: boolean | null;
  /** White label / 白标支持 */
  whiteLabel?: boolean | null;
}

/** Backend feature format (snake_case) / 后端特性格式 */
interface FeaturesSchemaRaw {
  advanced_analytics?: boolean | null;
  ai_enabled?: boolean | null;
  priority_support?: boolean | null;
  storage_billing_enabled?: boolean | null;
  white_label?: boolean | null;
}

/** Plan info (backend raw snake_case) / 套餐信息（后端原始） */
export interface TenantPlanInfoRaw {
  billing_cycle: BillingCycle;
  code: string;
  created_at: string;
  description?: null | string;
  features?: FeaturesSchemaRaw | null;
  id: number;
  is_active: boolean;
  name: string;
  price?: null | number | string;
  quota?: null | QuotaSchemaRaw;
  sort_order: number;
  updated_at?: string;
}

/** Plan info (frontend camelCase) / 套餐信息（前端） */
export interface TenantPlanInfo {
  billingCycle: BillingCycle;
  code: string;
  createdAt: string;
  description?: null | string;
  features?: FeaturesSchema | null;
  id: number;
  isActive: boolean;
  name: string;
  price?: null | number | string;
  quota?: null | QuotaSchema;
  sortOrder: number;
  updatedAt?: string;
}

/** Create plan request / 创建套餐请求 */
export interface TenantPlanCreateRequest {
  billing_cycle?: BillingCycle;
  code: string;
  description?: null | string;
  features?: FeaturesSchemaRaw | null;
  is_active?: boolean;
  name: string;
  price?: null | number | string;
  quota?: null | QuotaSchemaRaw;
  sort_order?: number;
}

/** Update plan request / 更新套餐请求 */
export interface TenantPlanUpdateRequest {
  billing_cycle?: BillingCycle | null;
  description?: null | string;
  features?: FeaturesSchemaRaw | null;
  is_active?: boolean | null;
  name?: null | string;
  price?: null | number | string;
  quota?: null | QuotaSchemaRaw;
  sort_order?: null | number;
}

/** Toggle status request / 切换状态请求 */
export interface TenantPlanStatusRequest {
  is_active: boolean;
}

/** Set permissions request / 设置权限请求 */
export interface TenantPlanPermissionsRequest {
  permission_ids: number[];
}

/** Permission info (tree structure) / 权限信息（树形） */
export interface PermissionInfo {
  children?: PermissionInfo[];
  code: string;
  id: number;
  name: string;
  parentId?: null | number;
}

/** Permission simple info (plan assigned permissions) / 权限简要信息 */
export interface PermissionSimpleInfo {
  code: string;
  id: number;
  name: string;
  resource: string;
  type: string;
}

/** Backend permission format / 后端权限格式 */
interface PermissionInfoRaw {
  children?: PermissionInfoRaw[];
  code: string;
  id: number;
  name: string;
  parent_id?: null | number;
}

/** Plan list query params / 套餐列表查询参数 */
export type TenantPlanListParams = Record<string, unknown>;

/** Paginated list response / 分页列表响应 */
export interface TenantPlanListResponse {
  items: TenantPlanInfo[];
  page: number;
  page_size: number;
  total: number;
}

/** Dropdown option (follows generic remote select pattern) / 下拉选项 */
export interface TenantPlanSelectOption {
  label: string;
  value: number;
  extra?: null | {
    billing_cycle?: string;
    code?: string;
  };
  disabled?: boolean;
  children?: null;
  is_leaf?: boolean | null;
}

// ============================================================
// Transform functions / 转换函数
// ============================================================

/** Transform quota snake_case -> camelCase / 转换配额 */
function transformQuota(raw?: null | QuotaSchemaRaw): null | QuotaSchema {
  if (!raw) return null;
  return {
    allowCustomDomain: raw.allow_custom_domain,
    apiCallsPerMonth: raw.api_calls_per_month,
    maxAdmins: raw.max_admins,
    maxCustomDomains: raw.max_custom_domains,
    maxFileSizeMb: raw.max_file_size_mb,
    maxUsers: raw.max_users,
    storageLimitGb: raw.storage_limit_gb,
  };
}

/** Transform features snake_case -> camelCase / 转换特性 */
function transformFeatures(
  raw?: FeaturesSchemaRaw | null,
): FeaturesSchema | null {
  if (!raw) return null;
  return {
    advancedAnalytics: raw.advanced_analytics,
    aiEnabled: raw.ai_enabled,
    prioritySupport: raw.priority_support,
    storageBillingEnabled: raw.storage_billing_enabled,
    whiteLabel: raw.white_label,
  };
}

/** Convert backend snake_case to frontend camelCase / 将后端转换为前端格式 */
function transformTenantPlanInfo(raw: TenantPlanInfoRaw): TenantPlanInfo {
  return {
    billingCycle: raw.billing_cycle,
    code: raw.code,
    createdAt: raw.created_at,
    description: raw.description,
    features: transformFeatures(raw.features),
    id: raw.id,
    isActive: raw.is_active,
    name: raw.name,
    price: raw.price,
    quota: transformQuota(raw.quota),
    sortOrder: raw.sort_order,
    updatedAt: raw.updated_at,
  };
}

/** Transform permission tree / 转换权限树 */
function transformPermission(raw: PermissionInfoRaw): PermissionInfo {
  return {
    children: raw.children?.map((child) => transformPermission(child)),
    code: raw.code,
    id: raw.id,
    name: raw.name,
    parentId: raw.parent_id,
  };
}

// ============================================================
// API functions / API 接口
// ============================================================

const API_PREFIX = '/admin/plans';

/**
 * Get plan list / 获取套餐列表
 * GET /admin/plans
 */
export async function getTenantPlanListApi(
  params?: TenantPlanListParams,
  options?: ApiRequestOptions,
): Promise<TenantPlanListResponse> {
  const response = await requestClient.get<{
    items: TenantPlanInfoRaw[];
    page: number;
    page_size: number;
    total: number;
  }>(API_PREFIX, { params, ...options });

  return {
    items: response.items.map((item) => transformTenantPlanInfo(item)),
    page: response.page,
    page_size: response.page_size,
    total: response.total,
  };
}

/**
 * Get plan detail / 获取套餐详情
 * GET /admin/plans/{plan_id}
 */
export async function getTenantPlanDetailApi(
  planId: number,
  options?: ApiRequestOptions,
): Promise<TenantPlanInfo> {
  const raw = await requestClient.get<TenantPlanInfoRaw>(
    `${API_PREFIX}/${planId}`,
    options,
  );
  return transformTenantPlanInfo(raw);
}

/**
 * Create plan / 创建套餐
 * POST /admin/plans
 */
export async function createTenantPlanApi(
  data: TenantPlanCreateRequest,
  options?: ApiRequestOptions,
): Promise<TenantPlanInfo> {
  const raw = await requestClient.post<TenantPlanInfoRaw>(
    API_PREFIX,
    data,
    options,
  );
  return transformTenantPlanInfo(raw);
}

/**
 * Update plan / 更新套餐
 * PUT /admin/plans/{plan_id}
 */
export async function updateTenantPlanApi(
  planId: number,
  data: TenantPlanUpdateRequest,
  options?: ApiRequestOptions,
): Promise<TenantPlanInfo> {
  const raw = await requestClient.put<TenantPlanInfoRaw>(
    `${API_PREFIX}/${planId}`,
    data,
    options,
  );
  return transformTenantPlanInfo(raw);
}

/**
 * Delete plan / 删除套餐
 * DELETE /admin/plans/{plan_id}
 */
export async function deleteTenantPlanApi(
  planId: number,
  options?: ApiRequestOptions,
): Promise<void> {
  await requestClient.delete(`${API_PREFIX}/${planId}`, options);
}

/**
 * Reorder plans / 重新排序套餐
 * PUT /admin/plans/reorder
 */
export async function reorderTenantPlansApi(
  ids: number[],
  options?: ApiRequestOptions,
): Promise<void> {
  await requestClient.put(`${API_PREFIX}/reorder`, { ids }, options);
}

/**
 * Toggle plan status / 切换套餐状态
 * PUT /admin/plans/{plan_id}/status
 */
export async function toggleTenantPlanStatusApi(
  planId: number,
  data: TenantPlanStatusRequest,
  options?: ApiRequestOptions,
): Promise<TenantPlanInfo> {
  // Use update API to toggle status / 使用 update 接口切换状态
  const raw = await requestClient.put<TenantPlanInfoRaw>(
    `${API_PREFIX}/${planId}`,
    data,
    options,
  );
  return transformTenantPlanInfo(raw);
}

/** Plan select option response / 套餐下拉选项响应 */
export interface TenantPlanSelectResponse {
  items: TenantPlanSelectOption[];
  total?: number;
  page?: number;
  page_size?: number;
  has_more?: boolean;
}

/**
 * Get plan select options / 获取套餐下拉选项
 * GET /admin/plans/select
 *
 * Response follows generic remote select pattern: { data: { items: [{label, value, extra}] } }
 */
export async function getTenantPlanSelectApi(params?: {
  is_active?: string;
  page?: number;
  page_size?: number;
  search?: string;
}): Promise<TenantPlanSelectResponse> {
  return requestClient.get<TenantPlanSelectResponse>(`${API_PREFIX}/select`, {
    params,
  });
}

/**
 * Get available permissions / 获取可分配的权限列表
 * GET /admin/plans/available-permissions
 */
export async function getAvailablePermissionsApi(
  options?: ApiRequestOptions,
): Promise<PermissionInfo[]> {
  const raw = await requestClient.get<PermissionInfoRaw[]>(
    `${API_PREFIX}/available-permissions`,
    options,
  );
  return raw.map((item) => transformPermission(item));
}

/**
 * Get plan permissions / 获取套餐权限
 * GET /admin/plans/{plan_id}/permissions
 *
 * Backend returns PermissionSimpleInfo[], frontend extracts id array for PermissionSelector
 */
export async function getTenantPlanPermissionsApi(
  planId: number,
  options?: ApiRequestOptions,
): Promise<PermissionSimpleInfo[]> {
  return requestClient.get<PermissionSimpleInfo[]>(
    `${API_PREFIX}/${planId}/permissions`,
    options,
  );
}

/**
 * Set plan permissions / 设置套餐权限
 * PUT /admin/plans/{plan_id}/permissions
 */
export async function setTenantPlanPermissionsApi(
  planId: number,
  data: TenantPlanPermissionsRequest,
  options?: ApiRequestOptions,
): Promise<void> {
  await requestClient.put(`${API_PREFIX}/${planId}/permissions`, data, {
    showSuccessMessage: true,
    ...options,
  });
}

File: backend/app/schemas/tenant/plan.py
"""
企业套餐相关 Schema / Tenant Plan Schema

定义套餐管理 API 的请求和响应数据结构
Defines plan management API request and response data structures.
"""

from collections.abc import Callable
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import Field, field_validator

from app.core.base_schema import BaseSchema
from app.core.i18n import _
from app.enums import BillingCycle

# ==================== 配额结构 / Quota Structure ====================


class QuotaSchema(BaseSchema):
    """配额结构定义 / Quota structure definition"""

    storage_limit_gb: int | None = Field(None, ge=0, description="存储限制(GB)")
    max_users: int | None = Field(None, ge=0, description="最大用户数")
    max_admins: int | None = Field(None, ge=0, description="最大管理员数")
    max_custom_domains: int | None = Field(None, ge=0, description="最大自定义域名数")
    allow_custom_domain: bool | None = Field(None, description="是否允许自定义域名")
    api_calls_per_month: int | None = Field(None, ge=0, description="每月API调用次数")
    max_file_size_mb: int | None = Field(None, ge=0, description="最大文件大小(MB)")

    def to_dict(self) -> dict[str, Any]:
        """转换为字典，仅包含非空值 / Convert to dict, non-null values only"""
        return {k: v for k, v in self.model_dump().items() if v is not None}


class FeaturesSchema(BaseSchema):
    """特性标记结构定义 / Features structure definition"""

    ai_enabled: bool | None = Field(None, description="是否启用AI功能")
    advanced_analytics: bool | None = Field(None, description="是否启用高级分析")
    white_label: bool | None = Field(None, description="是否支持白标")
    priority_support: bool | None = Field(None, description="是否优先支持")
    storage_billing_enabled: bool | None = Field(
        None, description="是否启用对象存储账单对账收费"
    )

    def to_dict(self) -> dict[str, Any]:
        """转换为字典，仅包含非空值 / Convert to dict, non-null values only"""
        return {k: v for k, v in self.model_dump().items() if v is not None}


# ==================== 响应 Schema / Response Schemas ====================


class TenantPlanResponse(BaseSchema):
    """套餐响应 / Plan response"""

    id: int = Field(..., description="套餐 ID")
    code: str = Field(..., description="套餐代码")
    name: str = Field(..., description="套餐名称")
    description: str | None = Field(None, description="套餐描述")
    price: Decimal | None = Field(None, description="价格")
    billing_cycle: str = Field(..., description="计费周期")
    is_active: bool = Field(..., description="是否启用")
    sort_order: int = Field(0, description="排序顺序")
    quota: dict | None = Field(None, description="配额配置")
    features: dict | None = Field(None, description="特性标记")
    tenants_count: int = Field(0, description="使用该套餐的企业数")
    permissions_count: int = Field(0, description="关联权限数")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    @classmethod
    def from_model(cls, plan) -> "TenantPlanResponse":
        """从模型创建响应 / Build response from model"""
        return cls(
            id=plan.id,
            code=plan.code,
            name=plan.name,
            description=plan.description,
            price=plan.price,
            billing_cycle=plan.billing_cycle,
            is_active=plan.is_active,
            sort_order=plan.sort_order,
            quota=plan.quota,
            features=plan.features,
            tenants_count=plan.tenants_count,
            permissions_count=plan.permissions_count,
            created_at=plan.created_at,
            updated_at=plan.updated_at,
        )


class PermissionSimpleResponse(BaseSchema):
    """权限简要响应（用于套餐详情） / Permission simple response (for plan detail)"""

    id: int = Field(..., description="权限 ID")
    code: str = Field(..., description="权限代码")
    name: str = Field(..., description="权限名称")
    type: str = Field(..., description="权限类型")
    resource: str = Field(..., description="资源标识")


class PermissionTreeSimpleResponse(BaseSchema):
    """权限树简要响应（用于套餐权限分配） / Permission tree simple response (for plan permission assignment)"""

    id: int = Field(..., description="权限 ID")
    code: str = Field(..., description="权限代码")
    name: str = Field(..., description="权限名称")
    type: str = Field(..., description="权限类型")
    resource: str | None = Field(None, description="资源标识")
    parent_id: int | None = Field(None, description="父级权限 ID")
    sort_order: int = Field(0, description="排序")
    children: list["PermissionTreeSimpleResponse"] = Field(
        default_factory=list, description="子权限"
    )


# 解决循环引用 / Resolve forward references
PermissionTreeSimpleResponse.model_rebuild()


class TenantPlanDetailResponse(TenantPlanResponse):
    """套餐详情响应（含权限列表） / Plan detail response (with permissions list)"""

    permissions: list[PermissionSimpleResponse] = Field(
        default_factory=list, description="关联权限列表"
    )

    @classmethod
    def from_model(
        cls,
        plan,
        translate_fn: Callable[[str], str] | None = None,
    ) -> "TenantPlanDetailResponse":
        """从模型创建详情响应 / Build detail response from model.

        Args:
            plan: TenantPlan ORM 对象 / TenantPlan ORM instance
            translate_fn: 可选的权限名称翻译函数 / Optional permission name translator
        """
        active_permissions = [p for p in plan.permissions if p.is_enabled]
        return cls(
            id=plan.id,
            code=plan.code,
            name=plan.name,
            description=plan.description,
            price=plan.price,
            billing_cycle=plan.billing_cycle,
            is_active=plan.is_active,
            sort_order=plan.sort_order,
            quota=plan.quota,
            features=plan.features,
            tenants_count=plan.tenants_count,
            permissions_count=len(active_permissions),
            created_at=plan.created_at,
            updated_at=plan.updated_at,
            permissions=[
                PermissionSimpleResponse(
                    id=p.id,
                    code=p.code,
                    name=translate_fn(p.name) if translate_fn else p.name,
                    type=p.type,
                    resource=p.resource,
                )
                for p in active_permissions
            ],
        )


# ==================== 请求 Schema ====================


class TenantPlanCreateRequest(BaseSchema):
    """创建套餐请求 / Create plan request"""

    name: str = Field(..., min_length=1, max_length=100, description="套餐名称")
    description: str | None = Field(None, max_length=500, description="套餐描述")
    price: Decimal | None = Field(None, ge=0, description="价格")
    billing_cycle: str = Field(
        default=BillingCycle.MONTHLY.value, description="计费周期"
    )
    is_active: bool = Field(True, description="是否启用")
    sort_order: int = Field(0, ge=0, description="排序顺序")
    quota: QuotaSchema | None = Field(None, description="配额配置")
    features: FeaturesSchema | None = Field(None, description="特性标记")

    @field_validator("billing_cycle")
    @classmethod
    def validate_billing_cycle(cls, v: str) -> str:
        """验证计费周期 / Validate billing cycle"""
        if v not in BillingCycle.values():
            raise ValueError(_("tenant_plan.invalid_billing_cycle"))
        return v


class TenantPlanUpdateRequest(BaseSchema):
    """更新套餐请求 / Update plan request"""

    name: str | None = Field(None, min_length=1, max_length=100, description="套餐名称")
    description: str | None = Field(None, max_length=500, description="套餐描述")
    price: Decimal | None = Field(None, ge=0, description="价格")
    billing_cycle: str | None = Field(None, description="计费周期")
    is_active: bool | None = Field(None, description="是否启用")
    sort_order: int | None = Field(None, ge=0, description="排序顺序")
    quota: QuotaSchema | None = Field(None, description="配额配置")
    features: FeaturesSchema | None = Field(None, description="特性标记")

    @field_validator("billing_cycle")
    @classmethod
    def validate_billing_cycle(cls, v: str | None) -> str | None:
        """验证计费周期 / Validate billing cycle"""
        if v is not None and v not in BillingCycle.values():
            raise ValueError(_("tenant_plan.invalid_billing_cycle"))
        return v


class TenantPlanPermissionsRequest(BaseSchema):
    """设置套餐权限请求 / Set plan permissions request"""

    permission_ids: list[int] = Field(
        default_factory=list, description="权限ID列表（仅支持菜单类型权限）"
    )


__all__ = [
    "QuotaSchema",
    "FeaturesSchema",
    "TenantPlanResponse",
    "TenantPlanDetailResponse",
    "PermissionSimpleResponse",
    "PermissionTreeSimpleResponse",
    "TenantPlanCreateRequest",
    "TenantPlanUpdateRequest",
    "TenantPlanPermissionsRequest",
]
